site/content/en/latest/user/security/private-key-provider.md

```diff
@@ -0,0 +1,385 @@
+  namespace: default
+spec:
+  targetRef:
+    group: gateway.networking.k8s.io
+    kind: Gateway
+    name: eg
+    namespace: default
+  type: JSONPatch
+  jsonPatches:
+    - type: "type.googleapis.com/envoy.extensions.transport_sockets.tls.v3.Secret"
+      name: default/example-cert
+      operation:
+        op: add
+        path: "/tls_certificate/private_key_provider"
+        value:
+          provider_name: qat
+          typed_config:
+            "@type": "type.googleapis.com/envoy.extensions.private_key_providers.qat.v3alpha.QatPrivateKeyMethodConfig"
+            private_key:
+              inline_string: |
+                abcd
+            poll_delay: 0.001s
+          fallback: true
+    - type: "type.googleapis.com/envoy.extensions.transport_sockets.tls.v3.Secret"
+      name: default/example-cert
+      operation:
+        op: copy
+        from: "/tls_certificate/private_key"
+        path: "/tls_certificate/private_key_provider/typed_config/private_key"
+EOF
+```
+
+### For CryptoMB
+
+```shell
+cat <<EOF | kubectl apply -f -
+apiVersion: gateway.envoyproxy.io/v1alpha1
+kind: EnvoyPatchPolicy
+metadata:
+  name: key-provider-patch-policy
+  namespace: default
+spec:
+  targetRef:
+    group: gateway.networking.k8s.io
+    kind: Gateway
+    name: eg
+    namespace: default
+  type: JSONPatch
+  jsonPatches:
+    - type: "type.googleapis.com/envoy.extensions.transport_sockets.tls.v3.Secret"
+      name: default/example-cert
+      operation:
+        op: add
+        path: "/tls_certificate/private_key_provider"
+        value:
+          provider_name: cryptomb
+          typed_config:  
+            "@type": "type.googleapis.com/envoy.extensions.private_key_providers.cryptomb.v3alpha.CryptoMbPrivateKeyMethodConfig"
+            private_key:
+              inline_string: |
+                abcd
+            poll_delay: 0.001s
+          fallback: true
+    - type: "type.googleapis.com/envoy.extensions.transport_sockets.tls.v3.Secret"
+      name: default/example-cert
+      operation:
+        op: copy
+        from: "/tls_certificate/private_key"
+        path: "/tls_certificate/private_key_provider/typed_config/private_key"
+EOF
+```
+
+### Benchmark after enabling private key provider
+
+Benchmark the gateway with fortio.
+
+```shell
+fortio load -c 64 -k -qps 0 -t 30s -keepalive=false https://www.example.com:${NODE_PORT}
+```
+
+You will see a performance boost after private key provider enabled. For example, you will get results as below.
+
+Without private key provider:
+
+```shell
+All done 43069 calls (plus 10 warmup) 6.966 ms avg, 1435.4 qps
+```
+
+With CryptoMB private key provider, the QPS is over 2 times than without private key provider.
+
+```shell
+All done 93983 calls (plus 128 warmup) 40.880 ms avg, 3130.5 qps
+```
+
+With QAT private key provider, the QPS is over 3 times than without private key provider
+
+```shell
+All done 134746 calls (plus 128 warmup) 28.505 ms avg, 4489.6 qps
+```
+
+## Testing
```

Comment:
shouldn't this be added before, near L276 ?

Reply:
good point, let me fix it, thanks!